janusgraph-berkeleyje/src/test/java/org/janusgraph/graphdb/BerkeleyTransactionTest.java

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb;
+
+import com.sleepycat.je.dbi.CursorImpl;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+import org.janusgraph.core.JanusGraph;
+import org.janusgraph.core.JanusGraphFactory;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+class BerkeleyTransactionTest {
+    static {
+        // Needs for ignore assert when cursor is closed during scan
+        ClassLoader.getSystemClassLoader().setClassAssertionStatus(CursorImpl.class.getName(), false);
+    }
+
+    @Test
+    void longRunningTxShouldBeRolledBacked(@TempDir File dir) throws InterruptedException {
+        JanusGraph graph = JanusGraphFactory.open("berkeleyje:" + dir.getAbsolutePath());
+
+        GraphTraversalSource traversal = graph.traversal();
+        for (int i = 0; i < 100_000; i++) {
```

Comment:
Do we really need to add 100,000 vertices? That takes ~15 sec on my machine, but the test failed for me without your changes even if just one vertex is added here.

Reply:
Potentially cursor scan may complete before we rollback tx and then future do not throw exception. I ran 10 tests with one vertex 7 failed, 1k vertices 6 failed, but 10k 0 failed

I mark it `@FlakyTest(invocationCount = 5, minSuccess = 1)` and change to 1k vertices

Reply:
If 10k are enough to reliably reproduce the problem, then I would prefer that as a value here or maybe 20k which should already reduce the runtime of this test to something in the range of 1-2 seconds which is acceptable in my opinion. (Use a value where you're relatively sure that the tests works deterministically.)
We should only use `FlakyTest` as a workaround for legacy tests that are for some reason flaky where we don't know yet why they are flaky and how we can change them to work deterministically and not as a feature for new tests.

Reply:
Ok, I set up 15k